### DISCUSSION TITLE
v8: Reactive dynamic route params

### DISCUSSION THREAD

**Discussion by jbmolle** [nuxt-modules/i18n#1588] [Q&A] [answered]
Hi,
Is it possible to have reactive dynamic route params?
Right now the doc shows:
```
<script setup>
definePageMeta({
  nuxtI18n: {
    en: { id: 'my-post' },
    fr: { id: 'mon-article' }
  }
})</script>
```
But 'my-post' and 'mon-article' are static strings. I'm getting the slugs from an external CMS. So would it be possible to do something like this:
```
<script setup>
const slugFr = ref('my-post')
slugFr.value = 'mon-article'

definePageMeta({
  nuxtI18n: {
    en: { id: 'my-post' },
    fr: { id: slugFr }
  }
})</script>
```
So I can get the slug after the api call at runtime.
Thanks,
JB

**Comment by jbmolle** [comment, marked as answer]
Hi,
I've found how to do it and I'm posting here so maybe it'll help other people.
In Nuxt 3 we can access the page meta through route.meta so:
```
<script setup>
const route = useRoute()
definePageMeta({
  nuxtI18n: {}
})
const slugFr = ref('initialValue')
watch(slugFr, (newSlug) => {
  route.meta.nuxtI18n = {
    en: { id: 'my-post' },
    fr: { id: newSlug }
  }
})
</script>
```
JB

**Reply by katerlouis** [reply]
@jbmolle Can you tell us how we do something similar in Nuxt2? `this.$route.meta` is `undefined` :( (atleast when you log it in `mounted()`)

**Reply by jbmolle** [reply]
Are you using nuxt-i18n v8? I'm not sure it works with Nuxt2.
If you're using nuxt-i18n v7 and vuex then you can do it following the doc here https://i18n.nuxtjs.org/lang-switcher#dynamic-route-parameters
I used to do it like this with Nuxt2 and it worked well.
```
<template>
  <!-- pages/post/_postId.vue -->
</template>

<script>
export default {
  async asyncData ({ store }) {
    await store.dispatch('i18n/setRouteParams', {
      en: { postId: 'my-post' },
      fr: { postId: 'mon-article' }
    })
  }
}
</script>

```

**Reply by katerlouis** [reply]
Thanks! I've found something similar in this storyblok specific article https://www.storyblok.com/tp/nuxtjs-translated-slugs, which seems to be exactly what I need, but I just can't get it to work... the crucial helper function `translatedSlugs` in the article is not documented and the source code is offline :(

It looks similar to your example, but I just can't get it to work `path` instead of `postId`! 

```js
      $store.dispatch('i18n/setRouteParams', {
        de: {
          path: '/test/de-123/',
        },
        en: {
          path: '/test/en-345/',
        },
      });
```

The good thing is: This code does change the routing; the "native path" (/impressum) is no longer working with this code, but I can't find the solution to get these routes working :o

**Reply by jbmolle** [reply]
Hi,
I'm sorry, I didn't understand you wanted to translate path too.
For the paths, from the documentation, you need to use (here in pages/about.vue):
```
export default {
  nuxtI18n: {
    paths: {
      en: '/about-us', // -> accessible at /about-us (no prefix since it's the default locale)
      fr: '/a-propos', // -> accessible at /fr/a-propos
      es: '/sobre'     // -> accessible at /es/sobre
    }
  }
}
```
And if you want it with dynamic route:
```
export default {
  nuxtI18n: {
    paths: {
      en: '/articles/:name',
      es: '/artículo/:name'
    }
  }
}
```

Can you give me the exact path transformation you want, please? What you give here is /test/de-123/ and /test/en-345/. Is there something after that like /test/de-123/my-article-in-german and /test/en-345/my-article-in-english?
If so I would create in nuxt pages the test directory, then a _category directory and then a _post.vue file.
And then you do:
```
$store.dispatch('i18n/setRouteParams', {
  de: {
    category: 'de-123',
    post: 'my-article-in-german'
  },
  en: {
    category: 'en-345',
    post: 'my-article-in-english'
  },
});
```
And if those categories are always de-123 and en-345, then you can create test directory -> de-123 directory -> then _post.vue file. And write:
```
export default {
  nuxtI18n: {
    paths: {
      de: '/test/de-123/:post',
      en: '/test/en-345/:post'
    }
  }
}

$store.dispatch('i18n/setRouteParams', {
  de: {
    post: 'my-article-in-german'
  },
  en: {
    post: 'my-article-in-english'
  },
});
```
Hope it helps, let me know if it's not what you're looking for.

**Reply by katerlouis** [reply]
I want to be available in browser (and reflected in cannocials, which is why I'm doing all of this)
`https://myhost.com/datenschutz/` (with slash)
and `https://myhost.com/en/privacy/` (with slash)

Thanks a lot for your help!

**Reply by jbmolle** [reply]
If you can edit your .nuxt/router.js, you can change path: '/datenschutz' with path: '/datenschutz//' after the routes are generated by nuxtI18n
But to do it directly in nuxtI18n, I'm not sure it's feasible as the / are probably removed

**Reply by jbmolle** [reply]
Or in your nuxt.config.ts:
router:{
  trailingSlash: true
}

**Reply by katerlouis** [reply]
The trailing slashes are not the problem; and the /datenschutz page is not in the router at all; its a dynamic page that is hosted on storyblok. In the router we use this asterisk route to funnel all storyblok pages:

```js
extendRoutes(routes, resolve) {
  routes.push({
    name: 'storyblok-funnel',
    path: '/*',
    component: resolve(__dirname, 'src/components/storyblok/story.vue'),
  });
},
```

The article is setting the path with `i18n/setRouteParams`, isn't it? There must be a way this is possible :D
Again: when I try doing this:

```js
      $store.dispatch('i18n/setRouteParams', {
        de: {
          path: '/datenschutz/',
        },
        en: {
          path: '/privacy/',
        },
      });
```

something is happening, but... I cant reach my pages anymore, neither in english nor in german.

**Reply by jbmolle** [reply]
Ah sorry I didn't understand that.
With your storyblok-funnel route * is pathMatch params in Nuxt
So can you try this, it should work:
```
$store.dispatch('i18n/setRouteParams', {
  de: {
    pathMatch: '/datenschutz/'
  },
  en: {
    pathMatch: '/privacy/'
  },
});
```

**Reply by katerlouis** [reply]
Thanks so much for your help! Unfortunately its still not working :'(
For whatever reason, `asyncData()`s `return`-statement is not reached when dispatching `i18n/setRouteParams` resulting in an empty page.

But the hassle doesn't seem to matter, ... because the canonicals don't get effected by new routeParams / new pathMatch– the issue remains the same. And I've tried all combinations of slashes for `pathMatch`

I tried looking at nuxt-i18ns source code if canonical links do even look at the store values; but its hard for me to track the code– on first glance, theres no mentioning of a store :(

```js
async asyncData({ $store, $storyblok, route, redirect, localePath }) {
  try {
    const story = await $storyblok.getStory(route.params.pathMatch); // pathMatch does not contain locale

    console.log('story is correctly fetched', story);

    $store.dispatch('i18n/setRouteParams', {
      de: {
        path: 'datenschutz', // just trying eeeverything
        pathMatch: 'datenschutz/', // tried /datenschutz, /datenschutz/, datenschutz/ and datenschutz without slashes...
      },
      en: {
        path: 'privacy',
        pathMatch: 'privacy/',
      },
    });

    console.log('this log is not happening...');

    return {
      story,
    };
  }

  catch (error) {
    console.error(error);
    // redirect(localePath({ name: '404' }));
  }
}
```

**Reply by jbmolle** [reply]
Sorry that it doesn't work.
You're sure that your store is defined? I don't think nuxt-i18n is going to create it if you haven't defined it. So in Nuxt you should at least create a store folder with an index.js file and a minimum store file, something like export const state = () => ({ counter: 0 })

**Reply by katerlouis** [reply]
First of all, thanks so much for your continuous support! I really appreciate the help and feel like we're getting close to a solution <3

Interesting that the store is not defined? That would explain why the console log after store.dispatch doesnt show up, I guess? But then again.. without doing the dispatch and never mentioning the i18n store in my code, the Vue devtools show the i18n store, so I guess it IS created after all? (see screenshot)

<img width="326" alt="image" src="https://user-images.githubusercontent.com/1983382/207331113-e4277b15-cd21-4bcc-9c84-820e3c8d913d.png">

Creating `src/store/i18n.js` with `export const state = () => ({ counter: 0 });` does result in the same error, the log after the dispatch isn't triggered (and therefore no return of the actual story content) – 

I mean.. if I must create the store myself, wouldn't I also need to create the proper state, mutations and the action `setRouteParams`?

EDIT: When I create an `src/store/i18n.js` I actually get an error saying: (see screenshot)
<img width="465" alt="image" src="https://user-images.githubusercontent.com/1983382/207347563-9eaadb58-e0e2-4d03-bb42-cdd3c9566ace.png">

**Reply by katerlouis** [reply]
I also tried different versions of `nuxt-i18n` – I was initially on `6.26.0` , now switched to `7.3.0` and still have the same issue.

**Reply by jbmolle** [reply]
Yes it seems the store is created so you don't have to. 
Do you have a testing repository that I could pull to try?

**Reply by katerlouis** [reply]
Unfortunately no; I will try to make a minimal reproduction of this issue! Thanks a lot– since it's only the two of us talking here, this could be considered spam, couldnt it? If you dont mind we could move this to a private conversation. I can offer Discord :)

**Reply by jbmolle** [reply]
Yes perfect, let's do Discord!
I will out of office for the next 2 weeks. I should be able to look at the code but if I don't, don't worry, I'll be back beginning of Jan

**Reply by katerlouis** [reply]
Thanks a lot! Here's my discord username, so you can add me :) katerlouis#7647
